### ISSUE TITLE
Supporting render-time variables

### ISSUE THREAD

**Issue by springmeyer** [mapnik/mapnik#2553]
This is a meta ticket for discussing and tracking progress on supporting render-time variables in Mapnik.

The idea of render-time variables is that any symbolizer can reference a variable like:

``` xml
<PolygonSymbolizer fill="@water" />
```

Or even like:

``` xml
<PolygonSymbolizer fill-opacity="[field_opacity]*@opacity" />
```

In CartoCSS this would be:

``` css
polygon-fill-opacity:"[field_opacity]*@opacity";
polygon-fill:"@water";
```

And then `{"water":"blue","opacity":.5}` could be passed at render time so that the value for `@water` or `@opacity` can be looked up dynamically when the symbolizer is processed. This allows the cartographer to selectively control what parts of the map should be dynamic and avoids the programmer needing to mutate/copy/template a mapnik.Map to enable tile-specific dynamism.

Currently if a variable exists in a stylesheet but its value is not available at render-time then no exception is thrown but rather a reasonable default value is used as a fallback: in the case of a color `transparent` is used (https://github.com/mapnik/mapnik/commit/aa85a4ff1eca4c5351b647c06f23d4b3e2a5d11b) and in the case of a number, 0 is used (division by zero is protected against).

In C++, as of Mapnik 3.x, all renderers (agg,cairo,svg, and grid) accept an optional hashmap of key:value pairs representing `variable name: variable value`.

The next step in making this support robust is:
- [ ] bindings in node-mapnik
  - [x] vtile.render: https://github.com/mapnik/node-mapnik/wiki/API#vectortilerendermap-surface-options-callback
  - [ ] map.render
- [ ] python bindings
- [ ]  regression/visual tests
- [ ] Develop a working convention on variable names that should be passed to Mapnik for different rendering scenarios. E.g. for tiled rendering scenarios variables like `zoom` and `scale-factor` would be frequently useful for cartographers to be able to depend upon and if all calling applications consistently used the same names this would make styles more interoperable between servers.

**Comment by knutole** [comment]
Will there be support for using variables as `filters`?  Like: 

```
[Area_m2/@Area_max > 0.5] {
       	polygon-fill: red;
    }
```

**Comment by springmeyer** [comment]
@knutole - yes, variables in filters will work.

**Comment by csytsma** [comment]
This would be a great feature, would love to see it implemented, especially for Zoom levels.  Any updates?

**Comment by hb0** [comment]
_Not sure if this is the same feature request but I give it a try:_
Is it possible to pass variables (e.g. in renderd.conf from mod_tile) to the mapnik.xml file?

_Sample of renderd.conf:_
`[mapnik_style1]
XML=/usr/local/share/maps/style/mapnik_style1.xml
UID=123`

_Sample of mapnik_style1.xml:_
`<Layer>
      <Datasource>
         <Parameter name="type"><![CDATA[postgis]]></Parameter>
         <Parameter name="table"><![CDATA[(SELECT wkb_geometry, id FROM geometries WHERE user_id = '@UID') as layer1]]></Parameter>`

**Comment by talaj** [comment]
> Is it possible to pass variables to the mapnik.xml file using the mapnik library (not node-mapnik)?

Yes it is. Variables are passed to the constructor of a renderer object from Mapnik library. In case of agg renderer, it's [this constructor](https://github.com/mapnik/mapnik/blob/743c14d3ff88e6c7d8b6ef647968a60d099cb762/include/mapnik/agg_renderer.hpp#L125). 
`renderd` has to use this constructor in order to support render-time variables.

**Comment by am2222** [comment]
@talaj Hi,
Is it possible to use this feature in python version? I could not find any clues about it

**Comment by talaj** [comment]
Hi @am2222 , there is [`render_with_vars()`](https://github.com/mapnik/python-mapnik/blob/eb849031e0fdcb2067f7596cb044e997a9e4dba7/src/mapnik_python.cpp#L829-L837), unfortunately with no tests.

**Comment by am2222** [comment]
@talaj  Hi, Thanks for your help, Actually I wanted to make mapnik get the color of fill parameter in polygon style from feature attribute I have no idea if the feature does this or not?

**Comment by talaj** [comment]
It is possible to set polygon color from feature attribute like this:

```xml
<PolygonSymbolizer fill="[color]" />
```

**Comment by am2222** [comment]
@talaj Hi,

I tried but it does not work, You know I am using MemoryDatasource and feed it using geojson strings in code,I first tried to define it using `style.fill='[color]'` but did not work.
Then I decided to write a simple xml map file and add the style as you said then attach that style in my code to the layer but it did not work. What other options do you think is possible?

**Comment by sommerluk** [comment]
Currently, (Mapnik 3.0.22) apparently `@zoom` is working fine when running a Mapnik style within Kosmtik. But it's not working fine when exporting from Kosmtik.

Is there any possibility to access SQL-like `!scale_denominator!` also in XML code?
